### PR TITLE
URL: Enhance the way long URLs are handled in filterURLForDisplay

### LINKED ISSUE #21100: [ MediaReplaceFlow ] Truncate the URL in the middle
**Is your feature request related to a problem? Please describe.**
When an image's replace menu is activated we can see a preview of the image's URL.
That URL is being truncated at the end.
The problem is that it is a bit useless when one can't see the filename because of all the other less important things like protocol and domain.

**Describe the solution you'd like**
The URL could be truncated from the middle so that we preserve the filenameappears as `localhost…Climbing-Wall.jpg` instead of `localhost:8888/wp-content/…` .

**Describe alternatives you've considered**
This is the way :)

### PR DESCRIPTION
**Pull request by hsingyuc**
## Description
Fix #21100
Add an argument to truncate the URL by max length.

## How has this been tested?
Additional tests were added to the `url/src/test/index.tests.js` file.

## Screenshots
truncated url

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by gziolo** [comment]
Thank you for your contribution. I like this proposal, it improves the experience when presenting the URL.

Could you refresh this branch because I see some CI checks failing. I hope it was only intermittent failures inherited from the main development branch. I see that all newly added unit tests pass so this should be good to go. By the way, excellent job covering new functionality with tests.

File: packages/url/README.md [hunk 1]
```diff
@@ -96,11 +96,13 @@ _Usage_
 
 ```js
 const displayUrl = filterURLForDisplay( 'https://www.wordpress.org/gutenberg/' ); // wordpress.org/gutenberg
+const imageUrl = filterURLForDisplay( 'https://www.wordpress.org/wp-content/uploads/img.png', 20 ); // …ent/uploads/img.png
 ```
 
 _Parameters_
 
 -   _url_ `string`: Original URL.
+-   _maxLength_ `(number|null)`: URL length.
```
Comment (gziolo) [inline packages/url/README.md:105]: This adds changes to the public API of `@wordpress/url` npm package. It is always a good idea to add an entry in the CHANGELOG so folks consuming this package could learn about new capabilities. How about we add a note in:

https://github.com/WordPress/gutenberg/blob/master/packages/url/CHANGELOG.md

Comment (hsingyuc) [reply packages/url/README.md:105]: @gziolo Thank you for the recommendation. Updated the changelog in https://github.com/WordPress/gutenberg/pull/27530/commits/1330576d52dc5302d5bfbdd675df12770b00b4d5 and rebased this PR.

Comment (gziolo) [reply packages/url/README.md:105]: Great, thank you!

**Review by gziolo — APPROVED** [review]
I must admit that the logic introduced is very complex, but the issue isn't simple as well. Fortunately, the tests added made me comfortable accepting this PR. 

Let's land this enhancement. Thank you 🎉

**Comment by kvf420** [comment]
This improvement should be extended to other file types. Currently, any long URL that isn't an image creates horizontal scrolling in the popover making edit button inaccessible.

Screenshot 2023-05-25 at 10 39 36 AM

**Pull request merged**